### PR TITLE
[FEAT] initial KV

### PR DESCRIPTION
**Pull request by aricart**

### CODE REVIEW

**Comment by aricart** [comment]
@ripienaar this is an initial thing, need to reimplement history to use heartbeats to bail early, and implement watch and status.

File: nats-base-client/kv.ts [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+export type Result = {
+  bucket: string;
+  key: string;
+  data: Uint8Array; // data lines up better with other nats apis in javascript
+  created: Date;
+  sequence: number;
+  delta?: number;
+  "origin_cluster"?: string;
+  operation: "PUT" | "DEL"; // these are redundant - a delete is empty payload...
```
Comment (ripienaar) [inline nats-base-client/kv.ts:44]: not really redundant though is it, you might legit choose to put a empty payload in a key.

File: nats-base-client/kv.ts [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+import { createInbox, headers, millis } from "./mod.ts";
+import { deferred, timeout } from "./util.ts";
+
+export type Result = {
+  bucket: string;
+  key: string;
+  data: Uint8Array; // data lines up better with other nats apis in javascript
+  created: Date;
+  sequence: number;
+  delta?: number;
```
Comment (ripienaar) [inline nats-base-client/kv.ts:42]: when is delta optional?

File: nats-base-client/kv.ts [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    this.js = js;
+    this.bucket = bucket;
+  }
+
+  async init(opts: Partial<BucketOpts> = {}): Promise<void> {
+    const sc = {} as StreamConfig;
+    sc.name = opts.streamName ?? this.bucketName();
+    sc.subjects = [this.subjectForBucket()];
+    sc.retention = RetentionPolicy.Limits;
+    this.maxPerSubject = sc.max_msgs_per_subject = opts.history ?? 5;
```
Comment (ripienaar) [inline nats-base-client/kv.ts:120]: above defaultBucketOps is history:1, should this align?

File: nats-base-client/kv.ts [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+  }
+
+  async get(k: string): Promise<Result> {
+    const sm = await this.jsm.streams.getMessage(this.bucketName(), {
+      last_by_subj: this.subjectForKey(k),
+    });
+    return this.toResult(k, sm);
+  }
+
+  async history(k: string): Promise<Result[]> {
```
Comment (ripienaar) [inline nats-base-client/kv.ts:203]: still not really sure how I want to expose history to clients so wasnt in the ADR iirc, but this seems fine anyway

**Review by ripienaar — COMMENTED** [review]
looks pretty good so far, presumably you confirmed interop with the CLI :)

**Comment by aricart** [comment]
> looks pretty good so far, presumably you confirmed interop with the CLI :)
next on the list :)

File: nats-base-client/kv.ts (hunk 2 above)
Comment (aricart) [reply nats-base-client/kv.ts:42]: I haven't mapped it.

File: nats-base-client/kv.ts (hunk 4 above)
Comment (aricart) [reply nats-base-client/kv.ts:203]: I want it for ngs

**Review by kozlovic — APPROVED** [review]
LGTM

**Comment by aricart** [comment]
@ripienaar merged this so I can use it from another project, but please do take a look at this.

**Comment by ripienaar** [comment]
had a look, good stuff, lets iterate based on your experience.  Will move the ADR into ADR repo and update jsm/cli with discussed things

**Pull request merged**